File: service/src/main/java/org/vsegda/util/TimeInstant.java
package org.vsegda.util;

import org.apache.commons.beanutils.Converter;

/**
 * @author Roman Elizarov
 */
public class TimeInstant {
    private final long time;
    private final TimePeriod period;
    private boolean nowOrFuture;

    private TimeInstant(long time, TimePeriod period) {
        this.time = time;
        this.period = period;
    }

    public static TimeInstant now() {
        return new TimeInstant(0, TimePeriod.ZERO);
    }

    public static TimeInstant valueOf(long time) {
        return new TimeInstant(time, null);
    }

    public static TimeInstant valueOf(TimePeriod period) {
        return new TimeInstant(0, period);
    }

    public static TimeInstant valueOf(String s) {
        if (s.isEmpty())
            return null;
        if (s.startsWith("+") || s.startsWith("-"))
            return new TimeInstant(0, TimePeriod.valueOf(s));
        else
            return new TimeInstant(TimeUtil.INSTANCE.parseTime(s), null);
    }

    public long time() {
        return period == null ? time : System.currentTimeMillis() + period.period();
    }

    public boolean isNowOrFuture() {
        return period == null ? time >= System.currentTimeMillis() : period.period() >= 0;
    }

    public TimeInstant subtract(TimePeriod other) {
        return period == null ?
                valueOf(time - other.period()) :
                valueOf(period.subtract(other));
    }

    public TimeInstant add(TimePeriod other) {
        return period == null ?
                valueOf(time + other.period()) :
                valueOf(period.add(other));
    }

    @Override
    public String toString() {
        return period == null ? TimeUtil.INSTANCE.formatDateTime(time) :
                ((period.period() >= 0 ? "+" : "") + period);
    }

    public static class Cnv implements Converter {
        public Object convert(Class clazz, Object o) {
            if (o == null)
               return null;
            String s = o.toString();
            if (s == null)
               return null;
            return valueOf(s);
        }
    }
}

File: service/src/main/java/org/vsegda/util/TimeInstant.kt
package org.vsegda.util

import org.apache.commons.beanutils.Converter

class TimeInstant private constructor(
    private val time: Long?,
    private val period: TimePeriod?
) {
    init { require(if (period == null) time != null else time == null) }

    private var _now: Long? = null
    private fun now(): Long = _now ?: System.currentTimeMillis().also { _now = it }
    
    val isNowOrFuture: Boolean
        get() = if (period == null) time!! >= now() else period.period >= 0

    fun time(): Long = if (period == null) time!! else now() + period.period

    operator fun plus(other: TimePeriod): TimeInstant = if (period == null)
        valueOf(time!! + other.period)
    else
        valueOf(period.minus(other))

    operator fun minus(other: TimePeriod): TimeInstant = if (period == null)
        valueOf(time!! - other.period)
    else
        valueOf(period.plus(other))

    override fun toString(): String = if (period == null)
        formatDateTime(time!!)
    else
        (if (period.period >= 0) "+" else "") + period

    class Cnv : Converter {
        @Suppress("UNCHECKED_CAST")
        override fun <T : Any?> convert(type: Class<T>, value: Any?): T {
            check(type === TimeInstant::class.java)
            return convert(value) as T
        }

        private fun convert(o: Any?): TimeInstant? {
            return valueOf(o?.toString() ?: return null)
        }
    }

    companion object {
        fun now(): TimeInstant = TimeInstant(null, TimePeriod.ZERO)
        fun valueOf(time: Long): TimeInstant = TimeInstant(time, null)
        fun valueOf(period: TimePeriod): TimeInstant = TimeInstant(null, period)

        fun valueOf(s: String): TimeInstant? {
            if (s.isEmpty()) return null
            return if (s.startsWith("+") || s.startsWith("-"))
                TimeInstant(null, TimePeriod.valueOf(s))
            else
                TimeInstant(parseTime(s), null)
        }
    }
}
